### PR TITLE
implemented basic functionality to add ports to k8s services

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/1197.

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -15,6 +15,6 @@ test:
 
 deployment:
   master:
-    branch: master
+    branch: svc
```
Comment (xh3b4sd) [inline circle.yml]: TODO

File: service/operator/service.go [hunk 2]
```diff
@@ -43,14 +99,147 @@ func New(config Config) (*Service, error) {
...
+
+func (s *Service) addFunc(obj interface{}) {
+	err := s.addFuncError(obj)
+	if err != nil {
+		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+	}
+}
+
+// TODO sync with mutex
+// TODO tests
```
Comment (xh3b4sd) [inline service/operator/service.go]: TODO

File: service/operator/service.go [hunk 3]
```diff
@@ -43,14 +99,147 @@ func New(config Config) (*Service, error) {
...
 }
+
+func (s *Service) addFunc(obj interface{}) {
+	err := s.addFuncError(obj)
+	if err != nil {
+		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+	}
+}
+
+// TODO sync with mutex
```
Comment (xh3b4sd) [inline service/operator/service.go]: TODO

Comment (xh3b4sd) [inline service/operator/service.go]: Done.

File: service/operator/service.go (hunk 2 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Done.

File: service/operator/service.go [hunk 4]
```diff
@@ -1,40 +1,118 @@
...
+	}
+
 	return Config{
 		// Dependencies.
-		Logger: nil,
+		K8sClient: k8sClient,
+		Logger:    newLogger,
+
+		// Settings.
+		Namespace: "default",
```
Comment (JosephSalisbury) [inline service/operator/service.go]: Configurable, please

File: service/operator/service.go [hunk 5]
```diff
@@ -1,40 +1,118 @@
...
+		Service:   "ingress-controller",
```
Comment (JosephSalisbury) [inline service/operator/service.go]: Same here, this should come from flags

File: service/operator/service.go [hunk 6]
```diff
@@ -1,40 +1,118 @@
...
 	}
 }
 
 // New creates a new configured service.
 func New(config Config) (*Service, error) {
 	// Dependencies.
+	if config.K8sClient == nil {
+		return nil, microerror.MaskAnyf(invalidConfigError, "config.K8sClient must not be empty")
```
Comment (JosephSalisbury) [inline service/operator/service.go]: We've tended to use just the struct field name, e.g: `K8sClient` up to now. I don't love `config.x` pattern

File: service/operator/service.go [hunk 7]
```diff
@@ -43,14 +121,152 @@ func New(config Config) (*Service, error) {
...
+	for _, p := range ports {
+		if p.Port == int32(number) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func portName(customObject kvmtpr.CustomObject) string {
```
Comment (JosephSalisbury) [inline service/operator/service.go]: should be in keys?

File: service/operator/service.go [hunk 8]
```diff
@@ -43,14 +121,152 @@ func New(config Config) (*Service, error) {
...
+	defer s.mutex.Unlock()
+
+	err := s.deleteFuncError(obj)
+	if err != nil {
+		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "delete")
+	}
+}
+
+func (s *Service) deleteFuncError(obj interface{}) error {
+	return nil
```
Comment (JosephSalisbury) [inline service/operator/service.go]: this is still to be implemented?

File: service/operator/service.go (hunk 6 above)
Comment (xh3b4sd) [inline service/operator/service.go]: I find this more descriptive and helpful for the developer hustling around the project.

File: service/operator/service.go (hunk 8 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Yes, in the next steps. Check the issue of the PR description. There is the most beautiful checklist somewhere in the middle. Next PRs are coming.

File: service/operator/service.go (hunk 7 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Good shout. Done.

File: service/operator/service.go (hunk 6 above)
Comment (JosephSalisbury) [inline service/operator/service.go]: Arguably, people will read this error message more than people writing it. No super strong feelings.

File: service/operator/service.go (hunk 4 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Done.

File: service/operator/service.go (hunk 5 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Done.

File: service/operator/service.go (hunk 6 above)
Comment (xh3b4sd) [inline service/operator/service.go]: Cool. I would like to keep it like this. Lets see if it turns out to be more useful than not.

**Comment by xh3b4sd** [comment]
Please check again.

File: circle.yml (hunk 1 above)
Comment (xh3b4sd) [inline circle.yml]: Done.

File: service/operator/service.go (hunk 6 above)
Comment (JosephSalisbury) [inline service/operator/service.go]: Works for me <3

**Comment by xh3b4sd** [comment]
I go with this now since Pawel has no time at the moment. We can get and apply feedback also later in case this is necessary. Merging now.

File: service/operator/service.go [hunk 9]
```diff
@@ -43,14 +121,161 @@ func New(config Config) (*Service, error) {
...
+
+	portName := PortName(customObject)
+
+	exists := portNameExists(k8sService.Spec.Ports, portName)
+	if exists {
+		s.logger.Log("debug", fmt.Sprintf("port for cluster '%s' already exists", ClusterID(customObject)))
+		return nil
+	}
+
+	clusterPort, err := s.newClusterPort(k8sService.Spec.Ports, ipsPortsToPorts(s.ipsPorts))
```
Comment (kopiczko) [inline service/operator/service.go]: I see `ipsPortsToPorts(s.ipsPorts) == availabePorts` Again I don't see where it's maintained. Am I missing something?

**Pull request merged**